### PR TITLE
Set THUMB bit for RUNTIME_FIXUP_HELPER functions

### PR DESCRIPTION
**Pull request by parjong**
This commit revises RUNTIME_FIXUP_HELPER invoke code as the second step toward resolving #6231.

This commit enforces THUMB bit to be set when RUNTIME_FIXUP_HELPER is called in order to fix ARM/THUMB inconsistency discussed in #6231.

### CODE REVIEW

File: src/zap/zapimport.cpp [hunk 1]
```diff
@@ -2165,6 +2165,11 @@ DWORD ZapIndirectHelperThunk::SaveWorker(ZapWriter * pZapWriter)
     *(WORD *)(p + 2) = 0xC000;
     p += 4;
 
+    // orr r12, r12, #1
```
Comment (jkotas) [inline src/zap/zapimport.cpp]: This should be fixed by filling in the indirection with the high bit set at runtime instead

Comment (parjong) [inline src/zap/zapimport.cpp]: @jkotas Thank you for comment. That approach is the first appropach that I tried, but I couldn't find where the related code is (I made this patch as a workaround). Could you let me know any information related with this? I read https://github.com/dotnet/coreclr/blob/master/Documentation/botr/readytorun-overview.md but couldn't find how it is really implemented in Core CLR.

Comment (jkotas) [inline src/zap/zapimport.cpp]: The related runtime code should be here: https://github.com/dotnet/coreclr/blob/master/src/vm/jitinterface.cpp#L13372

Comment (parjong) [inline src/zap/zapimport.cpp]: Thank you! I'll revise PR per feedback.

**Comment by parjong** [comment]
@jkotas I revised PR per feedback. Could you please take a look?

**Comment by jkotas** [comment]
I do not think this will work for things that are not real helpers (`READYTORUN_HELPER_Module` and `READYTORUN_HELPER_GSCookie`). Each of the values should be fixed up individually. We are using GetEEFuncEntryPoint macro everywhere else to get the right entrypoint. Can it be used here to fix the problem, something like:

```
    case READYTORUN_HELPER_DelayLoad_MethodCall:
        result = GetEEFuncEntryPoint(DelayLoad_MethodCall);
        break;
```

**Comment by parjong** [comment]
@jkotas I tried the following patch as you suggested, but it doesn't work for the tests of intreset:
```
diff --git a/src/vm/jitinterface.cpp b/src/vm/jitinterface.cpp
index 2343307..2d28a1b 100644
--- a/src/vm/jitinterface.cpp
+++ b/src/vm/jitinterface.cpp
@@ -13389,19 +13389,19 @@ BOOL LoadDynamicInfoEntry(Module *currentModule,
                     break;

                 case READYTORUN_HELPER_DelayLoad_MethodCall:
-                    result = (size_t)DelayLoad_MethodCall;
+                    result = (size_t)GetEEFuncEntryPoint(DelayLoad_MethodCall);
                     break;

                 case READYTORUN_HELPER_DelayLoad_Helper:
-                    result = (size_t)DelayLoad_Helper;
+                    result = (size_t)GetEEFuncEntryPoint(DelayLoad_Helper);
                     break;

                 case READYTORUN_HELPER_DelayLoad_Helper_Obj:
-                    result = (size_t)DelayLoad_Helper_Obj;
+                    result = (size_t)GetEEFuncEntryPoint(DelayLoad_Helper_Obj);
                     break;

                 case READYTORUN_HELPER_DelayLoad_Helper_ObjObj:
-                    result = (size_t)DelayLoad_Helper_ObjObj;
+                    result = (size_t)GetEEFuncEntryPoint(DelayLoad_Helper_ObjObj);
                     break;

                 default:
@@ -13410,9 +13410,6 @@ BOOL LoadDynamicInfoEntry(Module *currentModule,
                     return FALSE;
                 }
             }
-#ifdef _TARGET_ARM_
-            result |= THUMB_CODE;
-#endif
         }
         break;
```

I tried to get the definition of ``GetEEFuncEntryPoint``, and it seems that it is a simple macro that cast the variable to
```
src$ grep -Rn 'GetEEFuncEntryPoint'
...
vm/arm/cgencpu.h:40:#define GetEEFuncEntryPoint(pfn) 0x1001
vm/arm/cgencpu.h:42:#define GetEEFuncEntryPoint(pfn) GFN_TADDR(pfn)
...
src$ grep -Rn 'GFN_TADDR'
inc/daccess.h:1889:#define GFN_TADDR(name) (DacGlobalBase() + g_dacGlobals.fn__ ## name)
inc/daccess.h:2174:#define GFN_TADDR(name) ((TADDR)(name))
debug/daccess/daccess.cpp:5138:            if (inAddr != GFN_TADDR(DACNotifyCompilationFinished))
debug/daccess/daccess.cpp:5247:        *outAddr = GFN_TADDR(DACNotifyCompilationFinished);
debug/daccess/request.cpp:362:    threadpoolData->AsyncTimerCallbackCompletionFPtr = (CLRDATA_ADDRESS) GFN_TADDR(ThreadpoolMgr__AsyncTimerCallbackCompletion);
Binary file .vscode/.browse.VC.db matches
vm/virtualcallstub.cpp:1110:    if (GetIP(pContext) == GFN_TADDR(StubDispatchFixupPatchLabel)) {
vm/arm64/cgencpu.h:392:#define GetEEFuncEntryPoint(pfn) GFN_TADDR(pfn)
vm/amd64/cgencpu.h:159:#define GetEEFuncEntryPoint(pfn) GFN_TADDR(pfn)
vm/arm/cgencpu.h:42:#define GetEEFuncEntryPoint(pfn) GFN_TADDR(pfn)
vm/i386/cgencpu.h:51:#define GetEEFuncEntryPoint(pfn) GFN_TADDR(pfn)
```

Should I change ``GetEEFuncEntryPoint`` in ``arm/cgencpu.h`` to set THUMB_BIT?

**Comment by jkotas** [comment]
> Should I change `GetEEFuncEntryPoint` in arm/cgencpu.h to set THUMB_BIT?

Yes. I am surprised that nothing else is broken because of the `GetEEFuncEntryPoint` returning address without the thumb bit set.

**Comment by parjong** [comment]
@jkotas Oh, I could execute the tests without segfault when I changed GetEEFuncEntryPoint. I'll revise PR as suggested.

**Comment by jkotas** [comment]
There is one place in virtualcallstub.cpp that is calling `GFN_TADDR` directly. It should be changed to call `GetEEFuncEntryPoint` instead to get the thumb bit right (and save somebody time debugging it in future).

**Comment by parjong** [comment]
@dotnet-bot test Linx ARM Emulator Cross Debug Build please

**Comment by parjong** [comment]
@dotnet-bot test Linux ARM Emulator Cross Debug Build please

**Comment by jkotas** [comment]
Thanks!

**Pull request merged**